clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp

```diff
@@ -3068,11 +3107,217 @@ void MallocChecker::checkDeadSymbols(SymbolReaper &SymReaper,
+
+    const MemRegion *Base = RCall ? RCall : RExpr;
+    if (!Base) {
+      // Fallback: if we have a by-value record with unique_ptr fields but no
+      // region, mark all allocated symbols as escaped
+      ProgramStateRef State = C.getState();
+      RegionStateTy RS = State->get<RegionState>();
+      ProgramStateRef NewState = State;
+      for (auto [Sym, RefSt] : RS) {
+        if (RefSt.isAllocated() || RefSt.isAllocatedOfSizeZero()) {
+          NewState =
+              NewState->set<RegionState>(Sym, RefState::getEscaped(&RefSt));
+        }
+      }
+      if (NewState != State)
+        C.addTransition(NewState);
+      continue;
+    }
+
+    // Push direct smart owning pointer field regions only (precise root set).
+    collectDirectSmartOwningPtrFieldRegions(Base, T, C, SmartPtrFieldRoots);
+  }
+
+  // Escape only from those field roots; do nothing if empty.
+  if (!SmartPtrFieldRoots.empty()) {
+    ProgramStateRef State = C.getState();
+    auto Scan =
+        State->scanReachableSymbols<EscapeTrackedCallback>(SmartPtrFieldRoots);
+    ProgramStateRef NewState = Scan.getState();
+    if (NewState != State) {
+      C.addTransition(NewState);
+    } else {
+      // Fallback: if we have by-value record arguments but no smart pointer
+      // fields detected, check if any of the arguments are by-value records
+      // with smart pointer fields
+      bool hasByValueRecordWithSmartPtr = false;
+      for (unsigned I = 0, E = Call.getNumArgs(); I != E; ++I) {
+        const Expr *AE = Call.getArgExpr(I);
+        if (!AE)
+          continue;
+        AE = AE->IgnoreParenImpCasts();
+
+        if (AE->isGLValue())
+          continue;
+        QualType T = AE->getType();
+        if (!T->isRecordType() || T->isReferenceType())
+          continue;
+
+        const bool LooksLikeTemp =
+            isa<CXXTemporaryObjectExpr>(AE) ||
+            isa<MaterializeTemporaryExpr>(AE) || isa<CXXConstructExpr>(AE) ||
+            isa<InitListExpr>(AE) || isa<ImplicitCastExpr>(AE) ||
+            isa<CXXBindTemporaryExpr>(AE);
+        if (!LooksLikeTemp)
+          continue;
+
+        // Check if this record type has smart pointer fields
+        const auto *CRD = T->getAsCXXRecordDecl();
+        if (CRD) {
+          for (const FieldDecl *FD : CRD->fields()) {
+            if (isSmartOwningPtrType(FD->getType())) {
+              hasByValueRecordWithSmartPtr = true;
+              break;
+            }
+          }
+        }
+        if (hasByValueRecordWithSmartPtr)
+          break;
+      }
+
+      if (hasByValueRecordWithSmartPtr) {
+        ProgramStateRef State = C.getState();
+        RegionStateTy RS = State->get<RegionState>();
+        ProgramStateRef NewState = State;
+        for (auto [Sym, RefSt] : RS) {
+          if (RefSt.isAllocated() || RefSt.isAllocatedOfSizeZero()) {
+            NewState =
+                NewState->set<RegionState>(Sym, RefState::getEscaped(&RefSt));
+          }
+        }
```

Comment:
There is a lot of code duplication again and again.
Please think about organizing the code to a more maintainable form.

Reply:
The code was refactored and duplication was removed